File: leetcodes/206. Reverse Linked List/reverseLinkedList.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func reverseList(head *ListNode) *ListNode {
	//non or 1 node
	if head == nil || head.Next == nil {
		return head
	}
	//over 2 nodes
	var prev *ListNode
	for head != nil {
		temp :=head.Next
		head.Next = prev
		prev = head
		head = temp
	}
	return prev
}

File: leetcodes/206. Reverse Linked List/reverseLinkedList.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */
//iterative: time O(n), space O(1)
 function reverseList(head: ListNode | null): ListNode | null {
    let newHead = null;
    let temp = null;
    while(head != null) {
      temp = head.next
      head.next = newHead
      newHead = head
      head = temp
    }
    return newHead;
};
//recursive
function reverseList(head: ListNode | null): ListNode | null {
  return reverse(head, null)
}

function reverse(head: ListNode | null, newHead: ListNode |  null): ListNode | null {
  if (head == null) return newHead;
  let temp = head.next
  head.next = newHead
  return reverse(temp, head)
}
